File: benches/baselines/bubble_sort.c
// C baseline: bubble sort — compile with: gcc -O2 -o bubble bubble_sort.c
#include <stdio.h>
#include <time.h>

void bubble_sort(long arr[], int n) {
    for (int i = 0; i < n - 1; i++) {
        for (int j = 0; j < n - 1 - i; j++) {
            if (arr[j] > arr[j + 1]) {
                long tmp = arr[j];
                arr[j] = arr[j + 1];
                arr[j + 1] = tmp;
            }
        }
    }
}

int main(void) {
    struct timespec start, end;
    clock_gettime(CLOCK_MONOTONIC, &start);

    long result = 0;
    for (int iter = 0; iter < 1000; iter++) {
        long arr[] = {64, 34, 25, 12, 22, 11, 90, 1, 45, 78};
        bubble_sort(arr, 10);
        result = arr[0]; // 1 (sorted first element)
    }

    clock_gettime(CLOCK_MONOTONIC, &end);
    double elapsed = (end.tv_sec - start.tv_sec) + (end.tv_nsec - start.tv_nsec) / 1e9;
    printf("bubble_sort first = %ld, 1000 iterations in %.6f s (%.3f us/iter)\n",
           result, elapsed, elapsed * 1e6 / 1000);
    return 0;
}

File: benches/baselines/bubble_sort.rs
// Rust baseline: bubble sort — compile with: rustc -O -o bubble_rs bubble_sort.rs
use std::time::Instant;

fn bubble_sort(arr: &mut [i64]) {
    let n = arr.len();
    for i in 0..n.saturating_sub(1) {
        for j in 0..n - 1 - i {
            if arr[j] > arr[j + 1] {
                arr.swap(j, j + 1);
            }
        }
    }
}

fn main() {
    let start = Instant::now();
    let mut result = 0i64;
    for _ in 0..1000 {
        let mut arr = [64i64, 34, 25, 12, 22, 11, 90, 1, 45, 78];
        bubble_sort(&mut arr);
        result = arr[0];
    }
    let elapsed = start.elapsed();
    println!(
        "bubble_sort first = {result}, 1000 iterations in {:.6} s ({:.3} us/iter)",
        elapsed.as_secs_f64(),
        elapsed.as_secs_f64() * 1e6 / 1000.0
    );
}
